File: C/828-count-unique-characters-of-all-substrings-of-a-given-string/main.go
package main

import "fmt"

func uniqueLetterString(s string) int {
	ret := 0
	buf := make([][]int, 26)
	for i := 0; i < 26; i++ {
		buf[i] = []int{-1}
	}
	for i := 0; i < len(s); i++ {
		offset := byte(s[i]) - 'A'
		buf[offset] = append(buf[offset], i)
	}
	for i := 0; i < 26; i++ {
		buf[i] = append(buf[i], len(s))
	}

	for i := 0; i < 26; i++ {
		target := buf[i]
		if len(target) <= 2 {
			continue
		}

		for j := 1; j < len(target)-1; j++ {
			ret += (target[j] - target[j-1]) * (target[j+1] - target[j])
		}
	}

	return ret
}

func main() {
	fmt.Println(uniqueLetterString("LEETCODE"))
	fmt.Println(uniqueLetterString("ABC"))
	fmt.Println(uniqueLetterString("ABA"))
	fmt.Println(uniqueLetterString("AAAB"))
}

File: C/828-count-unique-characters-of-all-substrings-of-a-given-string/main.rs
struct Solution {}

impl Solution {
  pub fn unique_letter_string(s: String) -> i32 {
    let mut buf: Vec<Vec<i32>> = vec![Vec::new(); 26];

    (0..26).for_each(|idx| {
      buf[idx].push(-1);
    });
    s.as_bytes().iter().enumerate().for_each(|(i, &b)| {
      let offset = (b - b'A') as usize;
      buf[offset].push(i as i32);
    });

    (0..26).for_each(|idx| {
      buf[idx].push(s.len() as i32);
    });

    let mut ret: i32 = 0;
    (0..26).for_each(|idx| {
      if buf[idx].len() <= 2 {
        return;
      }
      let mut j: usize = 1;
      while j < buf[idx].len() - 1 {
        ret += (buf[idx][j] - buf[idx][j - 1]) * (buf[idx][j + 1] - buf[idx][j]);
        j += 1;
      }
    });
    ret
  }
}
